ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs

```diff
@@ -755,3 +825,423 @@ injectLedgerEvent index =
+-- >>> \ \\draw (0,0) rectangle (9,5);\
+-- >>> \ \\node (rect) at (1.5,4) [draw,minimum width=1cm,minimum height=0.5cm] {TxIn};\
+-- >>> \ \\node (oneOf) at (3.5,4) [draw=none] {NS};\
+-- >>> \ \\draw (rect) -> (oneOf);\
+-- >>> \ \\node (sh) at (6.5,4) [draw,minimum width=1cm,minimum height=0.5cm] {BlockATxOut};\
+-- >>> \ \\node (al) at (6.5,3) [draw,minimum width=1cm,minimum height=0.5cm] {BlockBTxOut};\
+-- >>> \ \\node (my) at (6.5,2) [draw=none,minimum width=1cm,minimum height=0.5cm] {...};\
+-- >>> \ \\node (ba) at (6.5,1) [draw,minimum width=1cm,minimum height=0.5cm] {BlockNTxOut};\
+-- >>> \ \\draw (oneOf) -> (sh);\
+-- >>> \ \\draw (oneOf) -> (al);\
+-- >>> \ \\draw (oneOf) -> (ba);\
+-- >>> \ \\draw (3,0.5) rectangle (8,4.5);"
+-- >>> :}
+--
+-- >>> :{
+-- >>> either (error . show) pure =<<
+-- >>>  renderToFile "docs/haddocks/hard-fork-tables-per-block.svg" defaultEnv (tikz ["positioning", "arrows"]) "\\node at (5,4.8) {\\small{LedgerState (HardForkBlock xs)}};\
+-- >>> \ \\draw (0,0) rectangle (10,5);\
+-- >>> \ \\node (oneOf2) at (2,4) [draw=none] {HardForkState};\
+-- >>> \ \\node (bb) at (5,4) [draw,minimum width=1cm,minimum height=0.5cm] {BlockAState};\
+-- >>> \ \\node (bt) at (8,4) [draw,minimum width=1cm,minimum height=0.5cm] {BlockATables};\
+-- >>> \ \\node (sb) at (5,3) [draw,minimum width=1cm,minimum height=0.5cm] {BlockBState};\
+-- >>> \ \\node (st) at (8,3) [draw,minimum width=1cm,minimum height=0.5cm] {BlockBTables};\
+-- >>> \ \\node (db) at (5,2) [draw=none,minimum width=1cm,minimum height=0.5cm] {...};\
+-- >>> \ \\node (dt) at (8,2) [draw=none,minimum width=1cm,minimum height=0.5cm] {...};\
+-- >>> \ \\node (bab) at (5,1) [draw,minimum width=1cm,minimum height=0.5cm] {BlockNState};\
+-- >>> \ \\node (bat) at (8,1) [draw,minimum width=1cm,minimum height=0.5cm] {BlockNTables};\
+-- >>> \ \\draw (oneOf2) -> (bb);\
+-- >>> \ \\draw (bb) -> (bt);\
+-- >>> \ \\draw (oneOf2) -> (sb);\
+-- >>> \ \\draw (sb) -> (st);\
+-- >>> \ \\draw (oneOf2) -> (bab);\
+-- >>> \ \\draw (bab) -> (bat);"
+-- >>> :}
+type DefaultHardForkTxOut xs = NS WrapTxOut xs
+
+class HasHardForkTxOut xs where
+  type HardForkTxOut xs :: Type
+  type HardForkTxOut xs = DefaultHardForkTxOut xs
+
+  injectHardForkTxOut :: Index xs x -> Value (LedgerState x) -> HardForkTxOut xs
+  distribHardForkTxOut :: Index xs x -> HardForkTxOut xs -> Value (LedgerState x)
+
+injectHardForkTxOutDefault ::
+     Index xs x
+  -> Value (LedgerState x)
+  -> DefaultHardForkTxOut xs
+injectHardForkTxOutDefault idx = injectNS idx . WrapTxOut
+
+distribHardForkTxOutDefault ::
+     CanHardFork xs
+  => Index xs x
+  -> DefaultHardForkTxOut xs
+  -> Value (LedgerState x)
+distribHardForkTxOutDefault idx  =
+    unwrapTxOut
+  . apFn (projectNP idx $ composeTxOutTranslations $ ipTranslateTxOut hardForkEraTranslation)
```

Comment:
I think this would be enough to make it a CAF?
```haskell
class HasHardForkTxOut xs where
  ...
  txOutEjections      :: NP (K (NS WrapTxOut xs) -.-> WrapTxOut) xs
  default txOutEjections :: CanHardFork xs => NP (K (NS WrapTxOut xs) -.-> WrapTxOut) xs
  txOutEjections = composeTxOutTranslations $ ipTranslateTxOut hardForkEraTranslation
```